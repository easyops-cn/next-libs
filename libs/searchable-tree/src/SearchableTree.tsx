import React, { useState, useRef, useEffect } from "react";
import { Tree, Input, Icon } from "antd";
import { TreeProps, AntTreeNodeProps } from "antd/lib/tree";
import { MenuIcon } from "@easyops/brick-types";
import { GeneralIcon } from "@libs/basic-components";

type TreeIcon = MenuIcon | React.ComponentType<React.SVGAttributes<SVGElement>>;

function isMenuIcon(icon: TreeIcon): icon is MenuIcon {
  return (icon as MenuIcon).lib !== undefined;
}

export interface SearchableTreeNodeProps extends AntTreeNodeProps {
  title?: string;
  icon?: TreeIcon;
  children?: SearchableTreeNodeProps[];
}

export interface SearchableTreeProps {
  selectedKeys?: string[];
  checkedKeys?: string[];
  expandedKeys?: string[];
  configProps?: Omit<
    TreeProps,
    "selectedKeys" | "checkedKeys" | "expandedKeys"
  >;
  dataSource: SearchableTreeNodeProps[];
  searchable?: boolean;
  onSelect?(selectedKeys: string[]): void;
  onCheck?(
    checkedKeys: string[] | { checked: string[]; halfChecked: string[] }
  ): void;
}

export function SearchableTree(props: SearchableTreeProps): React.ReactElement {
  const {
    selectedKeys: _selectedKeys,
    checkedKeys: _checkedKeys,
    expandedKeys: _expandedKeys,
    configProps = {},
    dataSource = [],
    searchable = false,
  } = props;
  const [selectedKeys, setSelectedKeys] = useState<string[]>();
  const [checkedKeys, setCheckedKeys] = useState<
    | string[]
    | {
        checked: string[];
        halfChecked: string[];
      }
  >();
  const [expandedKeys, setExpandedKeys] = useState<string[]>();
  const [searchValue, setSearchValue] = useState<string>();
  const nodeMatchedRef = useRef<boolean>(false);

  useEffect(() => {
    setSelectedKeys(_selectedKeys);
  }, [_selectedKeys]);
  useEffect(() => {
    setCheckedKeys(_checkedKeys);
  }, [_checkedKeys]);
  useEffect(() => {
    setExpandedKeys(_expandedKeys);
  }, [_expandedKeys]);

  let searchValueLength: number;

  if (searchValue) {
    searchValueLength = searchValue.length;
  }

  const getExpandedKeysBySearchValue = (
    nodes: SearchableTreeNodeProps[],
    searchValue: string,
    expandedKeys: string[]
  ) => {
    let isHit = false;

    nodes.forEach((node) => {
      if (node.children) {
        if (
          getExpandedKeysBySearchValue(node.children, searchValue, expandedKeys)
        ) {
          expandedKeys.push(node.key);
          isHit = true;
        }
      } else {
        if (node.title.includes(searchValue)) {
          expandedKeys.push(node.key);
          isHit = true;
        }
      }
    });

    return isHit;
  };

  const onChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const value = event.target.value;

    setSearchValue(value);
    nodeMatchedRef.current = false;

    if (value) {
      const expandedKeys: string[] = [];

      getExpandedKeysBySearchValue(dataSource, value, expandedKeys);
      setExpandedKeys(expandedKeys);
    }
  };

  const onSelect = (selectedKeys: string[]) => {
    setSelectedKeys(selectedKeys);

    if (props.onSelect) {
      props.onSelect(selectedKeys);
    }
  };

  const onCheck = (
    checkedKeys: string[] | { checked: string[]; halfChecked: string[] }
  ) => {
    setCheckedKeys(checkedKeys);

    if (props.onCheck) {
      props.onCheck(checkedKeys);
    }
  };

  const onExpand = (expandedKeys: string[]) => {
    setExpandedKeys(expandedKeys);
  };

  const getTreeNodes = (list: SearchableTreeNodeProps[]) => {
    return list.map((item) => {
      const { children, title: _title, icon: _icon, key, ...itemProps } = item;
      let title: React.ReactNode = _title;
      let icon: React.ReactNode;

      if (searchValue && !children) {
        const index = _title.indexOf(searchValue);

        if (index >= 0) {
          const beforeStr = _title.substring(0, index);
          const afterStr = _title.substring(searchValueLength + index);

          title = (
            <span
              ref={(() => {
                if (!nodeMatchedRef.current) {
                  nodeMatchedRef.current = true;

                  return (el: HTMLElement) => {
                    if (el) {
                      (
                        el.closest(".ant-tree-node-content-wrapper") || el
                      ).scrollIntoView();
                    }
                  };
                } else {
                  return null;
                }
              })()}
            >
              {beforeStr}
              <span style={{ color: "#0071eb", backgroundColor: "#e6f7ff" }}>
                {searchValue}
              </span>
              {afterStr}
            </span>
          );
        }
      }

      if (isMenuIcon(_icon)) {
        icon = <GeneralIcon icon={_icon} />;
      } else {
        icon = <Icon component={_icon} />;
      }

      return (
        <Tree.TreeNode {...itemProps} title={title} icon={icon} key={key}>
          {children && getTreeNodes(children)}
        </Tree.TreeNode>
      );
    });
  };

  return (
    <>
      {searchable && (
        <Input.Search
          onChange={onChange}
          style={{ marginBottom: 8 }}
          data-testid="search-input"
        />
      )}
      <div style={{ overflow: "auto" }}>
        <Tree
          {...configProps}
          selectedKeys={selectedKeys}
          checkedKeys={checkedKeys}
          expandedKeys={expandedKeys}
          onSelect={onSelect}
          onCheck={onCheck}
          onExpand={onExpand}
        >
          {getTreeNodes(dataSource)}
        </Tree>
      </div>
    </>
  );
}
