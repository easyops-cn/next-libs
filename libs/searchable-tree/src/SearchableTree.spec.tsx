import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { SearchableTree } from "./SearchableTree";
import { FaIcon } from "@next-core/brick-types";

const dataSource = [
  {
    title: "0",
    key: "0",
    icon: { lib: "fa", icon: "briefcase" } as FaIcon,
    children: [
      {
        title: "0-0",
        key: "00",
        icon: { lib: "fa", icon: "cube" } as FaIcon,
      },
      {
        title: "0-1",
        key: "01",
        icon: { lib: "fa", icon: "briefcase" } as FaIcon,
        children: [
          {
            title: "0-1-0",
            key: "010",
            icon: { lib: "fa", icon: "briefcase" } as FaIcon,
            children: [
              {
                title: "0-1-0-1",
                key: "0101",
                icon: { lib: "fa", icon: "cube" } as FaIcon,
              },
            ],
          },
        ],
      },
    ],
  },
  {
    title: "1",
    key: "1",
    icon: { lib: "fa", icon: "briefcase" } as FaIcon,
    children: [
      {
        title: "1-0",
        key: "10",
        icon: { lib: "fa", icon: "cube" } as FaIcon,
      },
    ],
  },
];

describe("SearchableTree", () => {
  it("should work", () => {
    const { asFragment } = render(<SearchableTree dataSource={dataSource} />);
    expect(asFragment()).toBeTruthy();
  });

  it("should expand the nodes that match search value", () => {
    const { asFragment, getByTestId, getByText, getAllByText } = render(
      <SearchableTree dataSource={dataSource} searchable />
    );

    fireEvent.change(getByTestId("search-input"), { target: { value: "1-0" } });
    expect(asFragment()).toBeTruthy();

    const mockScrollIntoView = getAllByText("1-0")[0].scrollIntoView;

    expect(mockScrollIntoView).toBeCalledTimes(1);
    fireEvent.click(getByText("0"));
    expect(mockScrollIntoView).toBeCalledTimes(1);
  });
});
