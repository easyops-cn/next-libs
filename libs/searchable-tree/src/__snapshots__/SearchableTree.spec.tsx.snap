// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchableTree should expand the nodes that match search value 1`] = `
<DocumentFragment>
  <span
    class="ant-input-search ant-input-affix-wrapper"
    style="margin-bottom: 8px;"
  >
    <input
      class="ant-input"
      data-testid="search-input"
      type="text"
      value="1-0"
    />
    <span
      class="ant-input-suffix"
    >
      <span
        aria-label="search"
        class="anticon anticon-search ant-input-search-icon"
        role="img"
        tabindex="-1"
      >
        <svg
          aria-hidden="true"
          class=""
          data-icon="search"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
          />
        </svg>
      </span>
    </span>
  </span>
  <div
    style="overflow: auto;"
  >
    <div
      class="ant-tree ant-tree-icon-hide"
    >
      <div
        role="tree"
      >
        <input
          style="width: 0px; height: 0px; display: flex; overflow: hidden; opacity: 0; border: 0px; padding: 0px; margin: 0px;"
          tabindex="0"
          value=""
        />
      </div>
      <div
        class="ant-tree-list"
        style="position: relative;"
      >
        <div
          class="ant-tree-list-holder"
        >
          <div>
            <div
              class="ant-tree-list-holder-inner"
              style="display: flex; flex-direction: column;"
            >
              <div
                class="ant-tree-treenode ant-tree-treenode-switcher-open"
              >
                <span
                  class="ant-tree-switcher ant-tree-switcher_open"
                >
                  <span
                    aria-label="caret-down"
                    class="anticon anticon-caret-down ant-tree-switcher-icon"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      class=""
                      data-icon="caret-down"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="0 0 1024 1024"
                      width="1em"
                    >
                      <path
                        d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                      />
                    </svg>
                  </span>
                </span>
                <span
                  class="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-open"
                  title="0"
                >
                  <span
                    class="ant-tree-title"
                  >
                    0
                  </span>
                </span>
              </div>
              <div
                class="ant-tree-treenode ant-tree-treenode-switcher-close"
              >
                <span
                  aria-hidden="true"
                  class="ant-tree-indent"
                >
                  <span
                    class="ant-tree-indent-unit ant-tree-indent-unit-start"
                  />
                </span>
                <span
                  class="ant-tree-switcher ant-tree-switcher-noop"
                />
                <span
                  class="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-normal"
                  title="0-0"
                >
                  <span
                    class="ant-tree-title"
                  >
                    0-0
                  </span>
                </span>
              </div>
              <div
                class="ant-tree-treenode ant-tree-treenode-switcher-open ant-tree-treenode-leaf-last"
              >
                <span
                  aria-hidden="true"
                  class="ant-tree-indent"
                >
                  <span
                    class="ant-tree-indent-unit ant-tree-indent-unit-end"
                  />
                </span>
                <span
                  class="ant-tree-switcher ant-tree-switcher_open"
                >
                  <span
                    aria-label="caret-down"
                    class="anticon anticon-caret-down ant-tree-switcher-icon"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      class=""
                      data-icon="caret-down"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="0 0 1024 1024"
                      width="1em"
                    >
                      <path
                        d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                      />
                    </svg>
                  </span>
                </span>
                <span
                  class="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-open"
                  title="0-1"
                >
                  <span
                    class="ant-tree-title"
                  >
                    0-1
                  </span>
                </span>
              </div>
              <div
                class="ant-tree-treenode ant-tree-treenode-switcher-open ant-tree-treenode-leaf-last"
              >
                <span
                  aria-hidden="true"
                  class="ant-tree-indent"
                >
                  <span
                    class="ant-tree-indent-unit ant-tree-indent-unit-end"
                  />
                  <span
                    class="ant-tree-indent-unit ant-tree-indent-unit-start ant-tree-indent-unit-end"
                  />
                </span>
                <span
                  class="ant-tree-switcher ant-tree-switcher_open"
                >
                  <span
                    aria-label="caret-down"
                    class="anticon anticon-caret-down ant-tree-switcher-icon"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      class=""
                      data-icon="caret-down"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="0 0 1024 1024"
                      width="1em"
                    >
                      <path
                        d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                      />
                    </svg>
                  </span>
                </span>
                <span
                  class="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-open"
                  title="0-1-0"
                >
                  <span
                    class="ant-tree-title"
                  >
                    0-1-0
                  </span>
                </span>
              </div>
              <div
                class="ant-tree-treenode ant-tree-treenode-switcher-open ant-tree-treenode-leaf-last"
              >
                <span
                  aria-hidden="true"
                  class="ant-tree-indent"
                >
                  <span
                    class="ant-tree-indent-unit ant-tree-indent-unit-end"
                  />
                  <span
                    class="ant-tree-indent-unit ant-tree-indent-unit-start ant-tree-indent-unit-end"
                  />
                  <span
                    class="ant-tree-indent-unit ant-tree-indent-unit-start ant-tree-indent-unit-end"
                  />
                </span>
                <span
                  class="ant-tree-switcher ant-tree-switcher-noop"
                />
                <span
                  class="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-normal"
                  title=""
                >
                  <span
                    class="ant-tree-title"
                  >
                    <span>
                      0-
                      <span
                        style="color: rgb(0, 113, 235); background-color: rgb(230, 247, 255);"
                      >
                        1-0
                      </span>
                      -1
                    </span>
                  </span>
                </span>
              </div>
              <div
                class="ant-tree-treenode ant-tree-treenode-switcher-open ant-tree-treenode-leaf-last"
              >
                <span
                  class="ant-tree-switcher ant-tree-switcher_open"
                >
                  <span
                    aria-label="caret-down"
                    class="anticon anticon-caret-down ant-tree-switcher-icon"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      class=""
                      data-icon="caret-down"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="0 0 1024 1024"
                      width="1em"
                    >
                      <path
                        d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                      />
                    </svg>
                  </span>
                </span>
                <span
                  class="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-open"
                  title="1"
                >
                  <span
                    class="ant-tree-title"
                  >
                    1
                  </span>
                </span>
              </div>
              <div
                class="ant-tree-treenode ant-tree-treenode-switcher-open ant-tree-treenode-leaf-last"
              >
                <span
                  aria-hidden="true"
                  class="ant-tree-indent"
                >
                  <span
                    class="ant-tree-indent-unit ant-tree-indent-unit-start ant-tree-indent-unit-end ant-tree-indent-unit-end-first-level"
                  />
                </span>
                <span
                  class="ant-tree-switcher ant-tree-switcher-noop"
                />
                <span
                  class="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-normal"
                  title=""
                >
                  <span
                    class="ant-tree-title"
                  >
                    <span>
                      <span
                        style="color: rgb(0, 113, 235); background-color: rgb(230, 247, 255);"
                      >
                        1-0
                      </span>
                    </span>
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`SearchableTree should work 1`] = `
<DocumentFragment>
  <div
    style="overflow: auto;"
  >
    <div
      class="ant-tree ant-tree-icon-hide"
    >
      <div
        role="tree"
      >
        <input
          style="width: 0px; height: 0px; display: flex; overflow: hidden; opacity: 0; border: 0px; padding: 0px; margin: 0px;"
          tabindex="0"
          value=""
        />
      </div>
      <div
        class="ant-tree-list"
        style="position: relative;"
      >
        <div
          class="ant-tree-list-holder"
        >
          <div>
            <div
              class="ant-tree-list-holder-inner"
              style="display: flex; flex-direction: column;"
            >
              <div
                class="ant-tree-treenode ant-tree-treenode-switcher-close"
              >
                <span
                  class="ant-tree-switcher ant-tree-switcher_close"
                >
                  <span
                    aria-label="caret-down"
                    class="anticon anticon-caret-down ant-tree-switcher-icon"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      class=""
                      data-icon="caret-down"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="0 0 1024 1024"
                      width="1em"
                    >
                      <path
                        d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                      />
                    </svg>
                  </span>
                </span>
                <span
                  class="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-close"
                  title="0"
                >
                  <span
                    class="ant-tree-title"
                  >
                    0
                  </span>
                </span>
              </div>
              <div
                class="ant-tree-treenode ant-tree-treenode-switcher-close ant-tree-treenode-leaf-last"
              >
                <span
                  class="ant-tree-switcher ant-tree-switcher_close"
                >
                  <span
                    aria-label="caret-down"
                    class="anticon anticon-caret-down ant-tree-switcher-icon"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      class=""
                      data-icon="caret-down"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="0 0 1024 1024"
                      width="1em"
                    >
                      <path
                        d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                      />
                    </svg>
                  </span>
                </span>
                <span
                  class="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-close"
                  title="1"
                >
                  <span
                    class="ant-tree-title"
                  >
                    1
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
